### ISSUE TITLE
MemcachedServerException (out of memory storing object) with memcached 1.5.12

### ISSUE THREAD

**Issue by ethanhere** [memcached/memcached#467]
I got the following exception when doing cas call on memcached 1.5 with java client:
2019-04-02 11:09:12,003 [ERROR] [pool-1-thread-1] com.nokia.solo.server.util.MemcachedCacheImpl - Exception occurred while performing cas operation on memcached: key=912417eb-c9e9-4b56-8f9e-f6e70f52b920 casId=1500790 errorMessage=MemcachedServerException: out of memory storing object,key=912417eb-c9e9-4b56-8f9e-f6e70f52b920
net.rubyeye.xmemcached.exception.MemcachedServerException: out of memory storing object,key=912417eb-c9e9-4b56-8f9e-f6e70f52b920
	at net.rubyeye.xmemcached.command.Command.decodeError(Command.java:262)
	at net.rubyeye.xmemcached.command.Command.decodeError(Command.java:279)
	at net.rubyeye.xmemcached.command.text.TextCASCommand.decode(TextCASCommand.java:83)
	at net.rubyeye.xmemcached.codec.MemcachedDecoder.decode0(MemcachedDecoder.java:61)
	at net.rubyeye.xmemcached.codec.MemcachedDecoder.decode(MemcachedDecoder.java:56)
	at com.google.code.yanf4j.nio.impl.NioTCPSession.decode(NioTCPSession.java:297)
	at com.google.code.yanf4j.nio.impl.NioTCPSession.decodeAndDispatch(NioTCPSession.java:237)
	at com.google.code.yanf4j.nio.impl.NioTCPSession.readFromBuffer(NioTCPSession.java:207)
	at com.google.code.yanf4j.nio.impl.AbstractNioSession.onRead(AbstractNioSession.java:196)
	at com.google.code.yanf4j.nio.impl.AbstractNioSession.onEvent(AbstractNioSession.java:341)
	at com.google.code.yanf4j.nio.impl.SocketChannelController.dispatchReadEvent(SocketChannelController.java:56)
	at com.google.code.yanf4j.nio.impl.NioController$ReadTask.run(NioController.java:110)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
FAILED CAS operation key=912417eb-c9e9-4b56-8f9e-f6e70f52b920 cas=1500790 valueUpdateSize=512000

The testing program opens 20 memcached clients, each of them creates 1 object with ramdom id and value. then get them by id and cas value, update (CAS) them with a large value (close to 500k length string). 6 out of those 20 cas calls failed with above pattern.

memcached information
start command
/usr/local/bin/memcached /usr/local/bin/memcached  -d -p 11211 -u nobody -m 1024 -c 1024 -P /var/run/memcached.pid -l 0.0.0.0

memcached console
[10:29:57 y1li:~/.ssh ] $ telnet 52.87.213.18 11211
Trying 52.87.213.18...
Connected to ec2-52-87-213-18.compute-1.amazonaws.com.
Escape character is '^]'.
version
VERSION 1.5.12
stats
STAT pid 8989
STAT uptime 5393
STAT time 1554218516
STAT version 1.5.12
STAT libevent 2.1.8-stable
STAT pointer_size 64
STAT rusage_user 49.557466
STAT rusage_system 38.332172
STAT max_connections 1024
STAT curr_connections 2
STAT total_connections 164
STAT rejected_connections 0
STAT connection_structures 23
STAT reserved_fds 20
STAT cmd_get 262
STAT cmd_set 1500005
STAT cmd_flush 0
STAT cmd_touch 0
STAT get_hits 262
STAT get_misses 0
STAT get_expired 0
STAT get_flushed 0
STAT delete_misses 0
STAT delete_hits 0
STAT incr_misses 0
STAT incr_hits 0
STAT decr_misses 0
STAT decr_hits 0
STAT cas_misses 0
STAT cas_hits 135
STAT cas_badval 0
STAT touch_hits 0
STAT touch_misses 0
STAT auth_cmds 0
STAT auth_errors 0
STAT bytes_read 2607152659
STAT bytes_written 12475662
STAT limit_maxbytes 1073741824
STAT accepting_conns 1
STAT listen_disabled_num 0
STAT time_in_listen_disabled_us 0
STAT threads 4
STAT conn_yields 0
STAT hash_power_level 19
STAT hash_bytes 4194304
STAT hash_is_expanding 0
STAT slab_reassign_rescues 2
STAT slab_reassign_chunk_rescues 0
STAT slab_reassign_evictions_nomem 9626
STAT slab_reassign_inline_reclaim 9623
STAT slab_reassign_busy_items 9626
STAT slab_reassign_busy_deletes 0
STAT slab_reassign_running 0
STAT slabs_moved 3
STAT lru_crawler_running 0
STAT lru_crawler_starts 3315
STAT lru_maintainer_juggles 3193473
STAT malloc_fails 0
STAT log_worker_dropped 0
STAT log_worker_written 0
STAT log_watcher_skipped 0
STAT log_watcher_sent 0
STAT bytes 953301162
STAT curr_items 555133
STAT total_items 1500007
STAT slab_global_page_pool 0
STAT expired_unfetched 0
STAT evicted_unfetched 935111
STAT evicted_active 0
STAT evictions 935111
STAT reclaimed 0
STAT crawler_reclaimed 0
STAT crawler_items_checked 5074661
STAT lrutail_reflocked 0
STAT moves_to_cold 1499870
STAT moves_to_warm 0
STAT moves_within_lru 0
STAT direct_reclaims 936403
STAT lru_bumps_dropped 0
END

**Comment by dormando** [comment]
Looks like you're low on RAM. What's your stats slabs/stats items?

SET/CAS of large items is challenging, since it allocates the memory ahead of time. if the larger slab classes have a low number of pages available it might not be able to evict anything else to make room.

The slab memory balancer tries to keep the tail age relatively even, but it looks like your test instance was up for an hourish and you're just flooding it with sets? Is that just to test it or is this realistic traffic?

**Comment by ethanhere** [comment]
I have a lot of free RAM:
[scbe@ip-10-255-1-87 ~]$ cat /proc/meminfo
MemTotal:        8057548 kB
MemFree:         6665092 kB
Buffers:           28388 kB
Cached:            58736 kB
SwapCached:            0 kB
Active:          1195448 kB
Inactive:          61056 kB
Active(anon):    1169392 kB
Inactive(anon):      132 kB
Active(file):      26056 kB
Inactive(file):    60924 kB
Unevictable:           0 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Dirty:                64 kB
Writeback:             0 kB
AnonPages:       1169412 kB
Mapped:            10584 kB
Shmem:               144 kB
Slab:              64688 kB
SReclaimable:       9844 kB
SUnreclaim:        54844 kB
KernelStack:        2496 kB
PageTables:         4788 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:     4028772 kB
Committed_AS:    1000760 kB
VmallocTotal:   34359738367 kB
VmallocUsed:       28288 kB
VmallocChunk:   34359708016 kB
HardwareCorrupted:     0 kB
AnonHugePages:   1122304 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
DirectMap4k:       12288 kB
DirectMap2M:     8376320 kB

here is stats slabs/stats items info:
[scbe@ip-10-255-1-87 ~]$ telnet localhost 11211
Trying ::1...
telnet: connect to address ::1: Connection refused
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
stats slabs
STAT 5:chunk_size 240
STAT 5:chunks_per_page 4369
STAT 5:total_pages 2
STAT 5:total_chunks 8738
STAT 5:used_chunks 8736
STAT 5:free_chunks 2
STAT 5:free_chunks_end 0
STAT 5:mem_requested 1979838
STAT 5:get_hits 4
STAT 5:cmd_set 13926
STAT 5:delete_hits 0
STAT 5:incr_hits 0
STAT 5:decr_hits 0
STAT 5:cas_hits 4
STAT 5:cas_badval 0
STAT 5:touch_hits 0
STAT 6:chunk_size 304
STAT 6:chunks_per_page 3449
STAT 6:total_pages 2
STAT 6:total_chunks 6898
STAT 6:used_chunks 6897
STAT 6:free_chunks 1
STAT 6:free_chunks_end 0
STAT 6:mem_requested 1880042
STAT 6:get_hits 11
STAT 6:cmd_set 32026
STAT 6:delete_hits 0
STAT 6:incr_hits 0
STAT 6:decr_hits 0
STAT 6:cas_hits 9
STAT 6:cas_badval 0
STAT 6:touch_hits 0
STAT 7:chunk_size 384
STAT 7:chunks_per_page 2730
STAT 7:total_pages 3
STAT 7:total_chunks 8190
STAT 7:used_chunks 8189
STAT 7:free_chunks 1
STAT 7:free_chunks_end 0
STAT 7:mem_requested 2823813
STAT 7:get_hits 19
STAT 7:cmd_set 40343
STAT 7:delete_hits 0
STAT 7:incr_hits 0
STAT 7:decr_hits 0
STAT 7:cas_hits 13
STAT 7:cas_badval 0
STAT 7:touch_hits 0
STAT 8:chunk_size 480
STAT 8:chunks_per_page 2184
STAT 8:total_pages 8
STAT 8:total_chunks 17472
STAT 8:used_chunks 17472
STAT 8:free_chunks 0
STAT 8:free_chunks_end 0
STAT 8:mem_requested 7551168
STAT 8:get_hits 20
STAT 8:cmd_set 48162
STAT 8:delete_hits 0
STAT 8:incr_hits 0
STAT 8:decr_hits 0
STAT 8:cas_hits 12
STAT 8:cas_badval 0
STAT 8:touch_hits 0
STAT 9:chunk_size 600
STAT 9:chunks_per_page 1747
STAT 9:total_pages 13
STAT 9:total_chunks 22711
STAT 9:used_chunks 22710
STAT 9:free_chunks 1
STAT 9:free_chunks_end 0
STAT 9:mem_requested 12284405
STAT 9:get_hits 21
STAT 9:cmd_set 60273
STAT 9:delete_hits 0
STAT 9:incr_hits 0
STAT 9:decr_hits 0
STAT 9:cas_hits 14
STAT 9:cas_badval 0
STAT 9:touch_hits 0
STAT 10:chunk_size 752
STAT 10:chunks_per_page 1394
STAT 10:total_pages 20
STAT 10:total_chunks 27880
STAT 10:used_chunks 27880
STAT 10:free_chunks 0
STAT 10:free_chunks_end 0
STAT 10:mem_requested 18867498
STAT 10:get_hits 36
STAT 10:cmd_set 75768
STAT 10:delete_hits 0
STAT 10:incr_hits 0
STAT 10:decr_hits 0
STAT 10:cas_hits 23
STAT 10:cas_badval 0
STAT 10:touch_hits 0
STAT 11:chunk_size 944
STAT 11:chunks_per_page 1110
STAT 11:total_pages 33
STAT 11:total_chunks 36630
STAT 11:used_chunks 36630
STAT 11:free_chunks 0
STAT 11:free_chunks_end 0
STAT 11:mem_requested 31082954
STAT 11:get_hits 43
STAT 11:cmd_set 96256
STAT 11:delete_hits 0
STAT 11:incr_hits 0
STAT 11:decr_hits 0
STAT 11:cas_hits 28
STAT 11:cas_badval 0
STAT 11:touch_hits 0
STAT 12:chunk_size 1184
STAT 12:chunks_per_page 885
STAT 12:total_pages 50
STAT 12:total_chunks 44250
STAT 12:used_chunks 44247
STAT 12:free_chunks 3
STAT 12:free_chunks_end 0
STAT 12:mem_requested 47091548
STAT 12:get_hits 52
STAT 12:cmd_set 120049
STAT 12:delete_hits 0
STAT 12:incr_hits 0
STAT 12:decr_hits 0
STAT 12:cas_hits 34
STAT 12:cas_badval 0
STAT 12:touch_hits 0
STAT 13:chunk_size 1480
STAT 13:chunks_per_page 708
STAT 13:total_pages 77
STAT 13:total_chunks 54516
STAT 13:used_chunks 54514
STAT 13:free_chunks 2
STAT 13:free_chunks_end 0
STAT 13:mem_requested 72653374
STAT 13:get_hits 63
STAT 13:cmd_set 147712
STAT 13:delete_hits 0
STAT 13:incr_hits 0
STAT 13:decr_hits 0
STAT 13:cas_hits 45
STAT 13:cas_badval 0
STAT 13:touch_hits 0
STAT 14:chunk_size 1856
STAT 14:chunks_per_page 564
STAT 14:total_pages 124
STAT 14:total_chunks 69936
STAT 14:used_chunks 69936
STAT 14:free_chunks 0
STAT 14:free_chunks_end 0
STAT 14:mem_requested 116659134
STAT 14:get_hits 73
STAT 14:cmd_set 188986
STAT 14:delete_hits 0
STAT 14:incr_hits 0
STAT 14:decr_hits 0
STAT 14:cas_hits 37
STAT 14:cas_badval 0
STAT 14:touch_hits 0
STAT 15:chunk_size 2320
STAT 15:chunks_per_page 451
STAT 15:total_pages 189
STAT 15:total_chunks 85239
STAT 15:used_chunks 85237
STAT 15:free_chunks 2
STAT 15:free_chunks_end 0
STAT 15:mem_requested 178070973
STAT 15:get_hits 89
STAT 15:cmd_set 231114
STAT 15:delete_hits 0
STAT 15:incr_hits 0
STAT 15:decr_hits 0
STAT 15:cas_hits 52
STAT 15:cas_badval 0
STAT 15:touch_hits 0
STAT 16:chunk_size 2904
STAT 16:chunks_per_page 361
STAT 16:total_pages 300
STAT 16:total_chunks 108300
STAT 16:used_chunks 108298
STAT 16:free_chunks 2
STAT 16:free_chunks_end 0
STAT 16:mem_requested 282931012
STAT 16:get_hits 100
STAT 16:cmd_set 292054
STAT 16:delete_hits 0
STAT 16:incr_hits 0
STAT 16:decr_hits 0
STAT 16:cas_hits 60
STAT 16:cas_badval 0
STAT 16:touch_hits 0
STAT 17:chunk_size 3632
STAT 17:chunks_per_page 288
STAT 17:total_pages 197
STAT 17:total_chunks 56736
STAT 17:used_chunks 56735
STAT 17:free_chunks 1
STAT 17:free_chunks_end 0
STAT 17:mem_requested 173516446
STAT 17:get_hits 57
STAT 17:cmd_set 153527
STAT 17:delete_hits 0
STAT 17:incr_hits 0
STAT 17:decr_hits 0
STAT 17:cas_hits 35
STAT 17:cas_badval 0
STAT 17:touch_hits 0
STAT 37:chunk_size 315872
STAT 37:chunks_per_page 3
STAT 37:total_pages 1
STAT 37:total_chunks 3
STAT 37:used_chunks 3
STAT 37:free_chunks 0
STAT 37:free_chunks_end 0
STAT 37:mem_requested 878686
STAT 37:get_hits 0
STAT 37:cmd_set 67
STAT 37:delete_hits 0
STAT 37:incr_hits 0
STAT 37:decr_hits 0
STAT 37:cas_hits 0
STAT 37:cas_badval 0
STAT 37:touch_hits 0
STAT 38:chunk_size 394840
STAT 38:chunks_per_page 2
STAT 38:total_pages 7
STAT 38:total_chunks 14
STAT 38:used_chunks 14
STAT 38:free_chunks 0
STAT 38:free_chunks_end 0
STAT 38:mem_requested 5394511
STAT 38:get_hits 0
STAT 38:cmd_set 299
STAT 38:delete_hits 0
STAT 38:incr_hits 0
STAT 38:decr_hits 0
STAT 38:cas_hits 0
STAT 38:cas_badval 0
STAT 38:touch_hits 0
STAT active_slabs 15
STAT total_malloced 1075838976
END
stats items
STAT items:5:number 8736
STAT items:5:number_hot 0
STAT items:5:number_warm 0
STAT items:5:number_cold 8736
STAT items:5:age_hot 0
STAT items:5:age_warm 0
STAT items:5:age 22450
STAT items:5:evicted 5186
STAT items:5:evicted_nonzero 5186
STAT items:5:evicted_time 12050
STAT items:5:outofmemory 0
STAT items:5:tailrepairs 0
STAT items:5:reclaimed 0
STAT items:5:expired_unfetched 0
STAT items:5:evicted_unfetched 5186
STAT items:5:evicted_active 0
STAT items:5:crawler_reclaimed 0
STAT items:5:crawler_items_checked 201332
STAT items:5:lrutail_reflocked 0
STAT items:5:moves_to_cold 13922
STAT items:5:moves_to_warm 0
STAT items:5:moves_within_lru 0
STAT items:5:direct_reclaims 5186
STAT items:5:hits_to_hot 4
STAT items:5:hits_to_warm 0
STAT items:5:hits_to_cold 0
STAT items:5:hits_to_temp 0
STAT items:6:number 6897
STAT items:6:number_hot 0
STAT items:6:number_warm 0
STAT items:6:number_cold 6897
STAT items:6:age_hot 0
STAT items:6:age_warm 0
STAT items:6:age 21789
STAT items:6:evicted 18223
STAT items:6:evicted_nonzero 18223
STAT items:6:evicted_time 18504
STAT items:6:outofmemory 0
STAT items:6:tailrepairs 0
STAT items:6:reclaimed 0
STAT items:6:expired_unfetched 0
STAT items:6:evicted_unfetched 18223
STAT items:6:evicted_active 0
STAT items:6:crawler_reclaimed 0
STAT items:6:crawler_items_checked 217475
STAT items:6:lrutail_reflocked 0
STAT items:6:moves_to_cold 32017
STAT items:6:moves_to_warm 0
STAT items:6:moves_within_lru 0
STAT items:6:direct_reclaims 18223
STAT items:6:hits_to_hot 11
STAT items:6:hits_to_warm 0
STAT items:6:hits_to_cold 0
STAT items:6:hits_to_temp 0
STAT items:7:number 8189
STAT items:7:number_hot 0
STAT items:7:number_warm 0
STAT items:7:number_cold 8189
STAT items:7:age_hot 0
STAT items:7:age_warm 0
STAT items:7:age 21872
STAT items:7:evicted 23953
STAT items:7:evicted_nonzero 23953
STAT items:7:evicted_time 18586
STAT items:7:outofmemory 0
STAT items:7:tailrepairs 0
STAT items:7:reclaimed 0
STAT items:7:expired_unfetched 0
STAT items:7:evicted_unfetched 23953
STAT items:7:evicted_active 0
STAT items:7:crawler_reclaimed 0
STAT items:7:crawler_items_checked 303154
STAT items:7:lrutail_reflocked 0
STAT items:7:moves_to_cold 40330
STAT items:7:moves_to_warm 0
STAT items:7:moves_within_lru 0
STAT items:7:direct_reclaims 23953
STAT items:7:hits_to_hot 19
STAT items:7:hits_to_warm 0
STAT items:7:hits_to_cold 0
STAT items:7:hits_to_temp 0
STAT items:8:number 17472
STAT items:8:number_hot 0
STAT items:8:number_warm 0
STAT items:8:number_cold 17472
STAT items:8:age_hot 0
STAT items:8:age_warm 0
STAT items:8:age 21911
STAT items:8:evicted 28495
STAT items:8:evicted_nonzero 28495
STAT items:8:evicted_time 14063
STAT items:8:outofmemory 0
STAT items:8:tailrepairs 0
STAT items:8:reclaimed 0
STAT items:8:expired_unfetched 0
STAT items:8:evicted_unfetched 28495
STAT items:8:evicted_active 0
STAT items:8:crawler_reclaimed 0
STAT items:8:crawler_items_checked 451143
STAT items:8:lrutail_reflocked 0
STAT items:8:moves_to_cold 48150
STAT items:8:moves_to_warm 0
STAT items:8:moves_within_lru 0
STAT items:8:direct_reclaims 28495
STAT items:8:hits_to_hot 20
STAT items:8:hits_to_warm 0
STAT items:8:hits_to_cold 0
STAT items:8:hits_to_temp 0
STAT items:9:number 22710
STAT items:9:number_hot 0
STAT items:9:number_warm 0
STAT items:9:number_cold 22710
STAT items:9:age_hot 0
STAT items:9:age_warm 0
STAT items:9:age 21845
STAT items:9:evicted 37549
STAT items:9:evicted_nonzero 37549
STAT items:9:evicted_time 18560
STAT items:9:outofmemory 0
STAT items:9:tailrepairs 0
STAT items:9:reclaimed 0
STAT items:9:expired_unfetched 0
STAT items:9:evicted_unfetched 37549
STAT items:9:evicted_active 0
STAT items:9:crawler_reclaimed 0
STAT items:9:crawler_items_checked 545416
STAT items:9:lrutail_reflocked 0
STAT items:9:moves_to_cold 60259
STAT items:9:moves_to_warm 0
STAT items:9:moves_within_lru 0
STAT items:9:direct_reclaims 37549
STAT items:9:hits_to_hot 21
STAT items:9:hits_to_warm 0
STAT items:9:hits_to_cold 0
STAT items:9:hits_to_temp 0
STAT items:10:number 27880
STAT items:10:number_hot 2
STAT items:10:number_warm 0
STAT items:10:number_cold 27878
STAT items:10:age_hot 3282
STAT items:10:age_warm 0
STAT items:10:age 21812
STAT items:10:evicted 47865
STAT items:10:evicted_nonzero 47865
STAT items:10:evicted_time 18531
STAT items:10:outofmemory 0
STAT items:10:tailrepairs 0
STAT items:10:reclaimed 0
STAT items:10:expired_unfetched 0
STAT items:10:evicted_unfetched 47865
STAT items:10:evicted_active 0
STAT items:10:crawler_reclaimed 0
STAT items:10:crawler_items_checked 670076
STAT items:10:lrutail_reflocked 0
STAT items:10:moves_to_cold 75743
STAT items:10:moves_to_warm 0
STAT items:10:moves_within_lru 0
STAT items:10:direct_reclaims 47865
STAT items:10:hits_to_hot 36
STAT items:10:hits_to_warm 0
STAT items:10:hits_to_cold 0
STAT items:10:hits_to_temp 0
STAT items:11:number 36630
STAT items:11:number_hot 2
STAT items:11:number_warm 0
STAT items:11:number_cold 36628
STAT items:11:age_hot 3285
STAT items:11:age_warm 0
STAT items:11:age 21852
STAT items:11:evicted 59598
STAT items:11:evicted_nonzero 59598
STAT items:11:evicted_time 18567
STAT items:11:outofmemory 0
STAT items:11:tailrepairs 0
STAT items:11:reclaimed 0
STAT items:11:expired_unfetched 0
STAT items:11:evicted_unfetched 59598
STAT items:11:evicted_active 0
STAT items:11:crawler_reclaimed 0
STAT items:11:crawler_items_checked 878546
STAT items:11:lrutail_reflocked 0
STAT items:11:moves_to_cold 96226
STAT items:11:moves_to_warm 0
STAT items:11:moves_within_lru 0
STAT items:11:direct_reclaims 59598
STAT items:11:hits_to_hot 43
STAT items:11:hits_to_warm 0
STAT items:11:hits_to_cold 0
STAT items:11:hits_to_temp 0
STAT items:12:number 44247
STAT items:12:number_hot 1
STAT items:12:number_warm 0
STAT items:12:number_cold 44246
STAT items:12:age_hot 3285
STAT items:12:age_warm 0
STAT items:12:age 21818
STAT items:12:evicted 75768
STAT items:12:evicted_nonzero 75768
STAT items:12:evicted_time 13967
STAT items:12:outofmemory 0
STAT items:12:tailrepairs 0
STAT items:12:reclaimed 0
STAT items:12:expired_unfetched 0
STAT items:12:evicted_unfetched 75768
STAT items:12:evicted_active 0
STAT items:12:crawler_reclaimed 0
STAT items:12:crawler_items_checked 1064256
STAT items:12:lrutail_reflocked 0
STAT items:12:moves_to_cold 120014
STAT items:12:moves_to_warm 0
STAT items:12:moves_within_lru 0
STAT items:12:direct_reclaims 75768
STAT items:12:hits_to_hot 52
STAT items:12:hits_to_warm 0
STAT items:12:hits_to_cold 0
STAT items:12:hits_to_temp 0
STAT items:13:number 54514
STAT items:13:number_hot 1
STAT items:13:number_warm 0
STAT items:13:number_cold 54513
STAT items:13:age_hot 3282
STAT items:13:age_warm 0
STAT items:13:age 21822
STAT items:13:evicted 93153
STAT items:13:evicted_nonzero 93153
STAT items:13:evicted_time 18540
STAT items:13:outofmemory 0
STAT items:13:tailrepairs 0
STAT items:13:reclaimed 0
STAT items:13:expired_unfetched 0
STAT items:13:evicted_unfetched 93153
STAT items:13:evicted_active 0
STAT items:13:crawler_reclaimed 0
STAT items:13:crawler_items_checked 1310699
STAT items:13:lrutail_reflocked 0
STAT items:13:moves_to_cold 147666
STAT items:13:moves_to_warm 0
STAT items:13:moves_within_lru 0
STAT items:13:direct_reclaims 93153
STAT items:13:hits_to_hot 63
STAT items:13:hits_to_warm 0
STAT items:13:hits_to_cold 0
STAT items:13:hits_to_temp 0
STAT items:14:number 69936
STAT items:14:number_hot 6
STAT items:14:number_warm 0
STAT items:14:number_cold 69930
STAT items:14:age_hot 3286
STAT items:14:age_warm 0
STAT items:14:age 21818
STAT items:14:evicted 119013
STAT items:14:evicted_nonzero 119013
STAT items:14:evicted_time 18537
STAT items:14:outofmemory 0
STAT items:14:tailrepairs 0
STAT items:14:reclaimed 0
STAT items:14:expired_unfetched 0
STAT items:14:evicted_unfetched 119013
STAT items:14:evicted_active 0
STAT items:14:crawler_reclaimed 0
STAT items:14:crawler_items_checked 1680849
STAT items:14:lrutail_reflocked 0
STAT items:14:moves_to_cold 188943
STAT items:14:moves_to_warm 0
STAT items:14:moves_within_lru 0
STAT items:14:direct_reclaims 119013
STAT items:14:hits_to_hot 73
STAT items:14:hits_to_warm 0
STAT items:14:hits_to_cold 0
STAT items:14:hits_to_temp 0
STAT items:15:number 85237
STAT items:15:number_hot 6
STAT items:15:number_warm 0
STAT items:15:number_cold 85231
STAT items:15:age_hot 3286
STAT items:15:age_warm 0
STAT items:15:age 21819
STAT items:15:evicted 145825
STAT items:15:evicted_nonzero 145825
STAT items:15:evicted_time 18537
STAT items:15:outofmemory 0
STAT items:15:tailrepairs 0
STAT items:15:reclaimed 0
STAT items:15:expired_unfetched 0
STAT items:15:evicted_unfetched 145825
STAT items:15:evicted_active 0
STAT items:15:crawler_reclaimed 0
STAT items:15:crawler_items_checked 2049504
STAT items:15:lrutail_reflocked 0
STAT items:15:moves_to_cold 231056
STAT items:15:moves_to_warm 0
STAT items:15:moves_within_lru 0
STAT items:15:direct_reclaims 145825
STAT items:15:hits_to_hot 89
STAT items:15:hits_to_warm 0
STAT items:15:hits_to_cold 0
STAT items:15:hits_to_temp 0
STAT items:16:number 108298
STAT items:16:number_hot 3
STAT items:16:number_warm 0
STAT items:16:number_cold 108295
STAT items:16:age_hot 3285
STAT items:16:age_warm 0
STAT items:16:age 21824
STAT items:16:evicted 183696
STAT items:16:evicted_nonzero 183696
STAT items:16:evicted_time 13976
STAT items:16:outofmemory 0
STAT items:16:tailrepairs 0
STAT items:16:reclaimed 0
STAT items:16:expired_unfetched 0
STAT items:16:evicted_unfetched 183696
STAT items:16:evicted_active 0
STAT items:16:crawler_reclaimed 0
STAT items:16:crawler_items_checked 2604064
STAT items:16:lrutail_reflocked 0
STAT items:16:moves_to_cold 291991
STAT items:16:moves_to_warm 0
STAT items:16:moves_within_lru 0
STAT items:16:direct_reclaims 183696
STAT items:16:hits_to_hot 100
STAT items:16:hits_to_warm 0
STAT items:16:hits_to_cold 0
STAT items:16:hits_to_temp 0
STAT items:17:number 56735
STAT items:17:number_hot 3
STAT items:17:number_warm 0
STAT items:17:number_cold 56732
STAT items:17:age_hot 3285
STAT items:17:age_warm 0
STAT items:17:age 21819
STAT items:17:evicted 96757
STAT items:17:evicted_nonzero 96757
STAT items:17:evicted_time 18536
STAT items:17:outofmemory 0
STAT items:17:tailrepairs 0
STAT items:17:reclaimed 0
STAT items:17:expired_unfetched 0
STAT items:17:evicted_unfetched 96757
STAT items:17:evicted_active 0
STAT items:17:crawler_reclaimed 0
STAT items:17:crawler_items_checked 1363839
STAT items:17:lrutail_reflocked 0
STAT items:17:moves_to_cold 153489
STAT items:17:moves_to_warm 0
STAT items:17:moves_within_lru 0
STAT items:17:direct_reclaims 96757
STAT items:17:hits_to_hot 57
STAT items:17:hits_to_warm 0
STAT items:17:hits_to_cold 0
STAT items:17:hits_to_temp 0
STAT items:37:number 3
STAT items:37:number_hot 0
STAT items:37:number_warm 0
STAT items:37:number_cold 3
STAT items:37:age_hot 0
STAT items:37:age_warm 0
STAT items:37:age 3280
STAT items:37:evicted 64
STAT items:37:evicted_nonzero 64
STAT items:37:evicted_time 0
STAT items:37:outofmemory 88
STAT items:37:tailrepairs 0
STAT items:37:reclaimed 0
STAT items:37:expired_unfetched 0
STAT items:37:evicted_unfetched 64
STAT items:37:evicted_active 0
STAT items:37:crawler_reclaimed 0
STAT items:37:crawler_items_checked 21
STAT items:37:lrutail_reflocked 0
STAT items:37:moves_to_cold 67
STAT items:37:moves_to_warm 0
STAT items:37:moves_within_lru 0
STAT items:37:direct_reclaims 961
STAT items:37:hits_to_hot 0
STAT items:37:hits_to_warm 0
STAT items:37:hits_to_cold 0
STAT items:37:hits_to_temp 0
STAT items:38:number 14
STAT items:38:number_hot 0
STAT items:38:number_warm 0
STAT items:38:number_cold 14
STAT items:38:age_hot 0
STAT items:38:age_warm 0
STAT items:38:age 10395
STAT items:38:evicted 285
STAT items:38:evicted_nonzero 285
STAT items:38:evicted_time 1
STAT items:38:outofmemory 134
STAT items:38:tailrepairs 0
STAT items:38:reclaimed 0
STAT items:38:expired_unfetched 0
STAT items:38:evicted_unfetched 285
STAT items:38:evicted_active 0
STAT items:38:crawler_reclaimed 0
STAT items:38:crawler_items_checked 176
STAT items:38:lrutail_reflocked 0
STAT items:38:moves_to_cold 299
STAT items:38:moves_to_warm 0
STAT items:38:moves_within_lru 0
STAT items:38:direct_reclaims 1649
STAT items:38:hits_to_hot 0
STAT items:38:hits_to_warm 0
STAT items:38:hits_to_cold 0
STAT items:38:hits_to_temp 0
END

In face my memcached instance has being running for about 1 day, I first fill it up with a lot of data so its size to be the max memory size (1GB), then I run the test described in the begining (20 thread, each of them to create an object, then update the object with big size (500k) value).

**Comment by dormando** [comment]
You don't have any free RAM within memcached, is what I meant. Your system memory doesn't have anything to do with this.

Memcached allocates the memory you allow it (1024 megabytes in this case) into slab classes separated by item size. You can see the "total_pages" in the slab stats output here, almost all of your memory are allocated to object sizes of 3k and less. You have 7 megabytes of memory allocates to the 512k object class. So you can have, at most, 14 objects being stored in parallel before you get an OOM.

If you run your test for long enough, the memory mover will reassign memory from the other slab classes into class 38 and your test will eventually pass. The memory mover is conservative since workloads don't tend to change drastically in a couple of seconds.

I'm not sure what you're trying to prove with your test? Why wait until it fills and then give it an aggressively different workload?

**Comment by ethanhere** [comment]
The test tries to mimic our prod scenario: we use memcached as cache server and it runs for a long time, definitely longer than a few days. So initially memcached would take request and create object by allocating more memory. Until to a certain time, the max memory is allocated, but the system would still take request to CRUD objects. We have seen the OOM error with CAS call with memcached 1.4. I created an issue (https://github.com/memcached/memcached/issues/455#issuecomment-462510042). You suggested to upgrade to 1.5. However we still see the same error with 1.5.

**Comment by dormando** [comment]
Do these 500k CAS writes happen all the time in production or only very rarely?

**Comment by ethanhere** [comment]
no sure how to define "rarely" but if a few (10 clients and 20 calls) happens together would cause this error.

In my test, when those 500k CAS calls, no client is using other objects, meaning the majority (over 90%) can be evicted to have the space. Do you have any idea how to tune it or trace more?

**Comment by dormando** [comment]
Yeah... so I guess you're sitting on an edge case. If the memory isn't typically used, there won't be anything in that slab class. If you have a sudden burst of writes all at once (and they're slow to actually write the data into the objects) you'll end up like this. You've actually seen this error happen in production under 1.5.12 or just in your testing?

The internal memory balancing algorithm has an equivalent external script:
https://github.com/memcached/memcached/blob/master/scripts/memcached-automove

In testing, you can just run the script against your test instance. It disables the internal algorithm while the script is running.

I should've put these as constants at the top of the script, but the > 2.5 and > 2 in here:
https://github.com/memcached/memcached/blob/master/scripts/memcached-automove#L84

You can crank these to 30.5 and 30 and re-run your test. The memory will have to get into those slab classes in the first place somehow, but it won't move the memory back out later. Some options:

1) Toy with the script, then either run the script alongside memcached or update the slabs_automove.c file a bit.
2) Disable automove entirely and manually configure RAM (a bit more work)
3) Run a separate instance that handles your large items. Probably simplest.

I have more options to improving the memory balancer these days but it's not a super quick change. :/

**Comment by dormando** [comment]
To reiterate for clarity: When uploading new data to an object, it has to first acquire memory to write into. There's no buffer while reading set from the network socket.

For example: If you have 7 megabytes of memory for class 512k, and are *simultaneously* uploading 14 objects, you be using 100% of the memory in class 512k for just pending uploads. The 15th request to SET will fail because there's nothing in memory that it can evict directly. Memory has to be moved from other slab classes, which happens slowly over time.

If the 512k class is only used temporarily, the memory mover sees no reason to keep memory in class 512k and reassigns it elsewhere.

**Comment by ethanhere** [comment]
I am new to the concept of automove. Can you explain that in more details:
About option 1
I see 2.5 in line 84 of that script, but not 2 in there. So I don't understand "but the > 2.5 and > 2 in here".
I don't understand how to "Toy with the script". Are you saying change 2.5 to 2 and run the script with testing memcached instance? Will the script adjust memcached instance's internal memory? After running that script, I run my test again?
 About option 2
How to "Disable automove entirely"? and how to "manually configure RAM"?

**Comment by dormando** [comment]
There's a `> 2` check about six lines below the 2.5 one.

I don't intend anything from this but trying to sync up so we can come to a solution: are you a frontend engineer or more of an operational person? I ask because we've not quite synced a few times and this script is fairly short. Also, can you please answer: `You've actually seen this error happen in production under 1.5.12 or just in your testing?` - it's important that I understand you've still seen the problem in production under 1.5.12 since your test is a bit contrived.

It sounds like you want to go with option 3 from my first response. If your "500k CAS" objects are coming from a specific part of your application, you should dedicate a specific memcached instance to them. Otherwise if you're only using the memory very rarely you'll keep ending up in this situation, and it sounds like you don't have a large pool of servers either.

If you want to stick with fewer instances and would like a more thorough audit of your setup, I'd ask that we set up a consulting job. Answering your questions is pretty specific to your situation and I can't dedicate the time to it, unfortunately :(

**Comment by ethanhere** [comment]
We have a prod system (using 1.4.4) that we have observed OOM when CAS a large(close to 500k) object. The test case we developed was intended to reproduce that OOM. So there is a difference between prod and the test case but we tried our best to be as close as possible. Since you recommended to upgrade to 1.5, we run the test case against 1.5 and still reproduced OOM.

In fact option 3 has problem to us. We cannot predict the size of update, therefore saving large object to a instance and saving smaller to another one will have a transaction problem, meaning deleting the object from one instance and adding to another instance.

Can you answer my question on option 1 and 2 in previous post?

**Comment by dormando** [comment]
I can't easily answer your questions. If you're still lost after what I've showed you it's going to require a significant time investment on my part. You should study the script yourself a bit and see if you can figure out what I was asking, or we can talk about a contract job to help you get up to speed.

I don't know what the difference is between your production and your test, so I can't recommend anything at this point. I'm sorry :(

**Comment by ethanhere** [comment]
I read this post (http://balodeamit.blogspot.com/2014/02/slab-reallocation-in-memcache.html), and it says that memory allocation could be of trouble if your keys changes slabs.
So I changed my code, when I caught MemecachedServerException OOM message, I delete and set. It seems that the problem could be solved that way. 
Can you comment and see if there is any problem with that approach?

**Comment by dormando** [comment]
It shouldn't fix your situation. I'm really sorry. I really want to help you but I don't know your background and don't know how to explain things to you :(

I can't help you anymore. I don't think you've tried 1.5.12 in production, you're just running tests to try to reproduce a fault. You should run 1.5.12 in production. Then, if *and only if* you continue to have trouble try the mailing list. Please do not continue to overthink this without running the newer version.

Github issues are only monitored by me. Maybe someone on the mailing list would have better luck helping you.

Thanks.

**Comment by dormando** [comment]
Sorry, I'm going to close this out. You should ask for help on the mailing list if you're still stuck.

**Comment by mark-i-m** [comment]
@dormando I'm experiencing an OOM with memcached 1.5.16. My workload is very contrived, but I would like to try understand what I'm seeing.

I have a 4TB in-memory memcached instance. My goal is to exercise the OS memory management implementation (even if the user-space workload is not very realistic). I started with a simple workload: I do SET operations with keys that are sequential integers (turned into strings) and values that are 512KB of zeros (to be precise, I am starting memcached with `-f 1.11` and inserting keys of size 523800 bytes, so there are 2 chunks per page and almost no memory wasted). These are the _only_ operations done on the server, and I am using only a single client thread.

memcached is given the entire 4TB of RAM. However, after inserting only 2TB of data, I get an OOM response from the server.

Any thoughts or insights would be helpful.

Here is the output of memcached tool:

```console
$ ./memcached-tool localhost
  #  Item_Size  Max_age   Pages   Count   Full?  Evicted Evict_Time OOM
 63   512.0K    153466s 2127649 4255298     yes        0        0    1

$ ./memcached-tool localhost stats
#localhost:11211   Field       Value
         accepting_conns           1
               auth_cmds           0
             auth_errors           0
                   bytes 2229204830958
              bytes_read 2229090461784
           bytes_written   102128520
              cas_badval           0
                cas_hits           0
              cas_misses           0
               cmd_flush           0
                 cmd_get           0
                 cmd_set     4255298
               cmd_touch           0
             conn_yields           0
   connection_structures           3
   crawler_items_checked   220678182
       crawler_reclaimed           0
        curr_connections           2
              curr_items     4255298
               decr_hits           0
             decr_misses           0
             delete_hits           0
           delete_misses           0
         direct_reclaims          10
          evicted_active           0
       evicted_unfetched           0
               evictions           0
       expired_unfetched           0
             get_expired           0
             get_flushed           0
                get_hits           0
              get_misses           0
              hash_bytes    33554432
       hash_is_expanding           0
        hash_power_level          22
               incr_hits           0
             incr_misses           0
                libevent 2.0.21-stable
          limit_maxbytes 4315368390656
     listen_disabled_num           0
        log_watcher_sent           0
     log_watcher_skipped           0
      log_worker_dropped           0
      log_worker_written           0
       lru_bumps_dropped           0
     lru_crawler_running           0
      lru_crawler_starts       17595
  lru_maintainer_juggles      830689
       lrutail_reflocked           0
            malloc_fails           0
         max_connections        1024
           moves_to_cold     4255298
           moves_to_warm           0
        moves_within_lru           0
                     pid       20873
            pointer_size          64
               reclaimed           0
    rejected_connections           0
            reserved_fds          20
           rusage_system 10764.436253
             rusage_user 8789.971169
   slab_global_page_pool           0
slab_reassign_busy_deletes           0
slab_reassign_busy_items           0
slab_reassign_chunk_rescues           0
slab_reassign_evictions_nomem           0
slab_reassign_inline_reclaim           0
   slab_reassign_rescues           0
   slab_reassign_running           0
             slabs_moved           0
                 threads           4
                    time  1564415652
time_in_listen_disabled_us           0
       total_connections           5
             total_items     4255298
              touch_hits           0
            touch_misses           0
                  uptime      153469
                 version      1.5.16

$ ./memcached-tool localhost settings
#localhost:11211   Field       Value
      auth_enabled_ascii          no
       auth_enabled_sasl          no
        binding_protocol auto-negotiate
             cas_enabled         yes
              chunk_size          48
          detail_enabled          no
           domain_socket        NULL
            dump_enabled         yes
               evictions         off
           flush_enabled         yes
           growth_factor        1.11
          hash_algorithm     murmur3
          hashpower_init           0
             hot_lru_pct          20
          hot_max_factor        0.20
            idle_timeout           0
   inline_ascii_response          no
                   inter        NULL
           item_size_max     1048576
             lru_crawler         yes
       lru_crawler_sleep         100
     lru_crawler_tocrawl           0
   lru_maintainer_thread         yes
           lru_segmented         yes
                maxbytes 4315368390656
                maxconns        1024
           maxconns_fast         yes
             num_threads           4
     num_threads_per_udp           4
                  oldest           0
          reqs_per_event          20
           slab_automove           1
     slab_automove_ratio        0.80
    slab_automove_window          30
          slab_chunk_max      524288
           slab_reassign         yes
         stat_key_prefix           :
        tail_repair_time           0
             tcp_backlog        1024
                 tcpport       11211
                temp_lru          no
           temporary_ttl          61
             track_sizes          no
                 udpport           0
                   umask         700
               verbosity           0
            warm_lru_pct          40
         warm_max_factor        2.00
     watcher_logbuf_size      262144
      worker_logbuf_size       65536
```

**Comment by dormando** [comment]
@mark-i-m can you open a new issue? It's unlikely your problem is related to this issue.

When you do, please include the output of `stats items` and `stats slabs` as well.
